ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -733,53 +761,94 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
     long totalCapacity = tgtLog.getCapacityInBytes();
     long writtenLastTime = 0;
     try (FileChannel fileChannel = Utils.openChannel(logSegmentToCopy.getView().getFirst(), false)) {
-      for (IndexEntry srcIndexEntry : srcIndexEntries) {
-        IndexValue srcValue = srcIndexEntry.getValue();
-        long usedCapacity = tgtIndex.getLogUsedCapacity();
-        if (isActive && (tgtLog.getCapacityInBytes() - usedCapacity >= srcValue.getSize())) {
-          fileChannel.position(srcValue.getOffset().getOffset());
-          Offset endOffsetOfLastMessage = tgtLog.getEndOffset();
-          // call into diskIOScheduler to make sure we can proceed (assuming it won't be 0).
-          diskIOScheduler.getSlice(DiskManager.CLEANUP_OPS_JOB_NAME, DiskManager.CLEANUP_OPS_JOB_NAME, writtenLastTime);
-          tgtLog.appendFrom(fileChannel, srcValue.getSize());
-          FileSpan fileSpan = tgtLog.getFileSpanForMessage(endOffsetOfLastMessage, srcValue.getSize());
-          if (srcValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
-            IndexValue putValue = tgtIndex.findKey(srcIndexEntry.getKey());
-            if (putValue != null) {
-              tgtIndex.markAsDeleted(srcIndexEntry.getKey(), fileSpan, srcValue.getOperationTimeInMs());
+      int startIndex = 0;
+      while (startIndex < srcIndexEntries.size()) {
+        // try to do a bundle of read to reduce disk IO
+        ByteBuffer bufferToUse;
+        long startOffset = srcIndexEntries.get(startIndex).getValue().getOffset().getOffset();
+        int endIndex;
+        if (bundleReadBuffer == null
+            || srcIndexEntries.get(startIndex).getValue().getSize() > bundleReadBuffer.capacity()) {
+          endIndex = startIndex;
+          bufferToUse = ByteBuffer.allocate((int) srcIndexEntries.get(startIndex).getValue().getSize());
+          srcMetrics.compactionBundleReadBufferNotFitIn.inc();
+          logger.trace("Record size greater than bundleReadBuffer capacity, key: {} size: {}",
+              srcIndexEntries.get(startIndex).getKey(), srcIndexEntries.get(startIndex).getValue().getSize());
+        } else {
+          endIndex = getBundleReadEndIndex(srcIndexEntries, startIndex);
+          bufferToUse = bundleReadBuffer;
+          bufferToUse.position(0);
+          bufferToUse.limit((int) (
+              srcIndexEntries.get(endIndex).getValue().getOffset().getOffset() + srcIndexEntries.get(endIndex)
+                  .getValue()
+                  .getSize() - startOffset));
+        }
+        // do IO read
+        int bytesRead = fileChannel.read(bufferToUse, startOffset);
+        if (bytesRead != bufferToUse.limit()) {
```

Comment:
What is this for? Are you checking if `bytesRead > bufferToUse.limit()`? That will never happen for two reasons. `ByteBuffer` does not allow a write that goes beyond the `limit` - an attempt to do that will lead to `BufferOverflowException`. Additionally, `FileChannel::read(ByteBuffer dst)` attempts to read at most `dst.remaining()` bytes into `dst`, as per its contract.